Make it possible to generate text previews for a particular HTML element

rr-codes opened the pull request:
#### 9784c8983ab995cac95a30e90acba5032ecfe6a0
<pre>
Make it possible to generate text previews for a particular HTML element
<a href="https://bugs.webkit.org/show_bug.cgi?id=281105">https://bugs.webkit.org/show_bug.cgi?id=281105</a>
<a href="https://rdar.apple.com/137559148">rdar://137559148</a>

Reviewed by Aditya Keerthi.

Expose functionality to be able to generate text previews or a UITargetedPreview for the text of a particular element, given its id.

This is a mutating operation, as the content has to temporarily be forced to be visible, and is then restored to its original state.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/StyledElement.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _targetedPreviewForElementWithID:completionHandler:]):
(-[WKWebView _textPreviewsForElementWithID:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextPreview.h: Copied from Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextPreview.mm: Copied from Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h.
(-[_WKTextPreview initWithSnapshotImage:presentationFrame:]):
(-[_WKTextPreview previewImage]):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createTextIndicatorForElementWithID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _createTargetedPreviewFromTextIndicator:previewContainer:]):
(-[WKContentView _createTargetedContextMenuHintPreviewIfPossible]):
(-[WKContentView targetedPreviewForID:completionHandler:]):
(-[WKContentView callCompletionHandlerForAnimationID:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::elementHasHiddenVisibility):
(WebKit::WebPage::createTextIndicatorForElementWithID):
(WebKit::WebPage::enableSourceTextAnimationAfterElementWithID): Deleted.
(WebKit::WebPage::enableTextAnimationTypeForElementWithID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/284942@main">https://commits.webkit.org/284942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/72ceb6b8999e13bb5cbd2afa7ab0cbfb8c7d1fd7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70989 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/50401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/22194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73105 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58200 "Failed to checkout and rebase branch from PR 34881") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/22012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/22194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74055 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/58200 "Failed to checkout and rebase branch from PR 34881") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/58200 "Failed to checkout and rebase branch from PR 34881") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/58200 "Failed to checkout and rebase branch from PR 34881") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/15221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/22012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/76809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/15265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10888 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/46202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/47274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/48557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/47016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextPreview.h:
```diff
@@ -0,0 +1,38 @@
+ */
+
+NS_SWIFT_SENDABLE
+WK_CLASS_AVAILABLE(macos(14.4), ios(17.4), visionos(1.1))
```
this is wrong

pxlcoder commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -44370,7 +44370,7 @@
 		F4CDF3CA27E9187200191928 /* Copy Profiling Data */ = {
 			isa = PBXShellScriptBuildPhase;
-			buildActionMask = 12;
+			buildActionMask = 2147483647;
```
Please undo.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4789,6 +4790,73 @@ - (void)_setDontResetTransientActivationAfterRunJavaScript:(BOOL)value
+            return;
+        }
+
+        RetainPtr contentView = strongSelf->_contentView;
```
This variable seems unnecessary.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -425,9 +426,11 @@ for this property.
 
-- (NSUUID *)_enableSourceTextAnimationAfterElementWithID:(NSString *)elementID WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
-- (NSUUID *)_enableFinalTextAnimationForElementWithID:(NSString *)elementID WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
-- (void)_disableTextAnimationWithUUID:(NSUUID *)nsUUID WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
```
I think we need to keep these around still?

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -18511,7 +18517,7 @@
 		2E16B6F42019BC25008996D6 /* Copy Additional Resources */ = {
 			isa = PBXShellScriptBuildPhase;
-			buildActionMask = 12;
+			buildActionMask = 2147483647;
```
Please undo.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -1044,6 +1044,76 @@
+        return;
+    }
+
+    styledElement->setInlineStyleProperty(CSSPropertyVisibility, initialVisibility, IsImportant::Yes);
```
This isn't really idempotent, because you could be adding inline style to an element that never had one.

rr-codes replied:
what should I do instead?

pxlcoder replied:
`removeInlineStyleProperty` if it never had one?

rr-codes committed:
Make it possible to generate text previews for a particular HTML element
https://bugs.webkit.org/show_bug.cgi?id=281105
rdar://137559148

Reviewed by Aditya Keerthi.

Expose functionality to be able to generate text previews or a UITargetedPreview for the text of a particular element, given its id.

This is a mutating operation, as the content has to temporarily be forced to be visible, and is then restored to its original state.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/StyledElement.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _targetedPreviewForElementWithID:completionHandler:]):
(-[WKWebView _textPreviewsForElementWithID:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextPreview.h: Copied from Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextPreview.mm: Copied from Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h.
(-[_WKTextPreview initWithSnapshotImage:presentationFrame:]):
(-[_WKTextPreview previewImage]):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createTextIndicatorForElementWithID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _createTargetedPreviewFromTextIndicator:previewContainer:]):
(-[WKContentView _createTargetedContextMenuHintPreviewIfPossible]):
(-[WKContentView targetedPreviewForID:completionHandler:]):
(-[WKContentView callCompletionHandlerForAnimationID:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::elementHasHiddenVisibility):
(WebKit::WebPage::createTextIndicatorForElementWithID):
(WebKit::WebPage::enableSourceTextAnimationAfterElementWithID): Deleted.
(WebKit::WebPage::enableTextAnimationTypeForElementWithID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/284942@main
